flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java

```diff
@@ -240,7 +243,7 @@ private boolean hasRegisteredState() {
 	}
 
 	@Override
-	public <N, SV, S extends State, IS extends S> IS createState(
+	public <N, SV, S extends State, IS extends S> IS createInternalState(
```

Comment:
Why are we adding `Internal` here? I would suggest to call the method `create(Internal?)KeyValueState`, because there will also be other state in the future (timers).

Reply:
The method uses state descriptor to produce backend specific `InternalKvState`s. Is it also applicable for timers? It could be named `createInternalKvState` and `InternalKvStateFactory`.

Reply:
No, timers cannot use state descriptor, they cannot extend `State`